src/main/java/ua/danit/final_project/configuration/security/JwtTokenUtil.java

```diff
@@ -0,0 +1,122 @@
+  private Clock clock = DefaultClock.INSTANCE;
+
+  @Value("${jwt.secret}")
+  private String secret;
+
+  @Value("${jwt.expiration}")
+  private Long expiration;
+
+  public String getUsernameFromToken(String token) {
+    return getClaimFromToken(token, Claims::getSubject);
+  }
+
+  public Date getIssuedAtDateFromToken(String token) {
+    return getClaimFromToken(token, Claims::getIssuedAt);
+  }
+
+  public Date getExpirationDateFromToken(String token) {
+    return getClaimFromToken(token, Claims::getExpiration);
+  }
+
+  public <T> T getClaimFromToken(String token, Function<Claims, T> claimsResolver) {
+    final Claims claims = getAllClaimsFromToken(token);
+    return claimsResolver.apply(claims);
+  }
+
+  private Claims getAllClaimsFromToken(String token) {
+    return Jwts.parser()
+        .setSigningKey(secret)
+        .parseClaimsJws(token)
+        .getBody();
+  }
+
+  private Boolean isTokenExpired(String token) {
+    final Date expiration = getExpirationDateFromToken(token);
+    return expiration.before(clock.now());
+  }
+
+  private Boolean isCreatedBeforeLastPasswordReset(Date created, Date lastPasswordReset) {
+    return (lastPasswordReset != null && created.before(lastPasswordReset));
+  }
+
+  private Boolean ignoreTokenExpiration(String token) {
+    // here you specify tokens, for that the expiration is ignored
+    return false;
+  }
+
+  public String generateToken(UserDetails userDetails) {
+    Map<String, Object> claims = new HashMap<>();
+    return doGenerateToken(claims, userDetails.getUsername());
+  }
+
+  private String doGenerateToken(Map<String, Object> claims, String subject) {
+    final Date createdDate = clock.now();
+    final Date expirationDate = calculateExpirationDate(createdDate);
+
+    return Jwts.builder()
+        .setClaims(claims)
+        .setSubject(subject)
+        .setIssuedAt(createdDate)
+        .setExpiration(expirationDate)
+        .signWith(SignatureAlgorithm.HS512, secret)
+        .compact();
+  }
+
+  public Boolean canTokenBeRefreshed(String token, Date lastPasswordReset) {
+    final Date created = getIssuedAtDateFromToken(token);
+    return !isCreatedBeforeLastPasswordReset(created, lastPasswordReset)
+        && (!isTokenExpired(token) || ignoreTokenExpiration(token));
+  }
+
+  public String refreshToken(String token) {
+    final Date createdDate = clock.now();
+    final Date expirationDate = calculateExpirationDate(createdDate);
+
+    final Claims claims = getAllClaimsFromToken(token);
+    claims.setIssuedAt(createdDate);
+    claims.setExpiration(expirationDate);
+
+    return Jwts.builder()
+        .setClaims(claims)
+        .signWith(SignatureAlgorithm.HS512, secret)
+        .compact();
+  }
+
+  public Boolean validateToken(String token, UserDetails userDetails) {
+    User user = (User) userDetails;
+    final String username = getUsernameFromToken(token);
+    final Date created = getIssuedAtDateFromToken(token);
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this useless assignment to local variable "created". ([squid:S1854](https://sonarcloud.io/organizations/final-project/rules#rule_key=squid%3AS1854))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=final-project&pullRequest=173&issues=AWb1IgB1W3Y69bz5KWFw&open=AWb1IgB1W3Y69bz5KWFw)

[ISSUE_KEY]: AWb1IgB1W3Y69bz5KWFw